### PR TITLE
Rename TypeVars to something more novice-friendly

### PR DESCRIPTION
**Pull request by uri-granta**
To make the function signatures less confusing for beginners, I've renamed nearly all of TypeVars to something more descriptive:

* `SearchSpaceType` (a type var bound to SearchSpace)
* `ProbabilisticModelType` (a type var bound to ProbabilisticModels)
* `TrainableProbabilisticModelType` (a type var bound to TrainableProbabilisticModel)
* `GIBBONModelType` (a type var bound to SupportsCovarianceObservationNoise used by GIBBON)
* `ResultType` (unbound type var used by AcquisitionRule)
* `StateType` (unbound type var used by Records and the BO loop)

Note that I've not included whether the types are contravariant in the type name as is sometimes done (e.g. `ProbabilisticModelType_contra`) as that makes them unweildy and is meaningless to beginners. However, I make sure to mention this in the type var docstrings.

### CODE REVIEW

File: trieste/acquisition/function/function.py [hunk 1]
```diff
@@ -36,7 +36,10 @@
     SingleModelVectorizedAcquisitionBuilder,
 )
 
-M_contra = TypeVar("M_contra", bound=ProbabilisticModel, contravariant=True)
+ProbabilisticModelType = TypeVar(
+    "ProbabilisticModelType", bound=ProbabilisticModel, contravariant=True
+)
+""" Contravariant type variable bound to :class:`~trieste.models.ProbabilisticModel`. """
```
Comment (henrymoss) [inline trieste/acquisition/function/function.py]: Could they all be defined somewhere else once?

File: trieste/acquisition/function/entropy.py [hunk 2]
```diff
@@ -228,10 +231,13 @@ class SupportsCovarianceObservationNoiseTrajectory(
     has an associated trajectory sampler."""
 
 
-GIBBONT = TypeVar("GIBBONT", bound=SupportsCovarianceObservationNoise, contravariant=True)
+GIBBONModelType = TypeVar(
+    "GIBBONModelType", bound=SupportsCovarianceObservationNoise, contravariant=True
+)
+""" Type variable bound to :class:`~trieste.models.SupportsCovarianceObservationNoise`. """
```
Comment (henrymoss) [inline trieste/acquisition/function/entropy.py:232]: Why do we not have one of these in the cts Thompson sampling acq? That needs the model to have a sampler

Comment (uri-granta) [reply trieste/acquisition/function/entropy.py:232]: In general we don't need these just to specify that you need a specific feature: e.g. GreedyContinuousThompsonSampling is defined as a SingleModelGreedyAcquisitionBuilder[HasTrajectorySampler]. You only need type variables if the features you require depend on one of the input arguments.

For example, a rule that gets passed an acqusition function builder that only supports models with predict_joint also only supports models with predict_joint. Or an acquisition function (such as GIBBON) that gets passed a sampler that only supports models with trajectory functions also only supports models with trajectory functions.

File: trieste/acquisition/function/function.py (hunk 1 above)
Comment (uri-granta) [reply trieste/acquisition/function/function.py]: Good suggestion. Done.

File: trieste/models/interfaces.py [hunk 3]
```diff
@@ -25,7 +25,10 @@
 from ..types import TensorType
 from ..utils import DEFAULTS
 
-T = TypeVar("T", bound="ProbabilisticModel", contravariant=True)
+ProbabilisticModelType = TypeVar(
```
Comment (hstojic) [inline trieste/models/interfaces.py:28]: should this be exported in __init__

File: trieste/utils/__init__.py [hunk 4]
```diff
@@ -14,7 +14,7 @@
 """ This package contains library utilities. """
 from typing import Callable, Mapping, TypeVar
 
-from .misc import DEFAULTS, Err, Ok, Result, T_co, jit, shapes_equal, to_numpy
+from .misc import DEFAULTS, Err, Ok, Result, ResultType, jit, shapes_equal, to_numpy
 
 K = TypeVar("K")
```
Comment (hstojic) [inline trieste/utils/__init__.py]: a small digression: why is map_values defined in __init__ and not in e.g. misc.py?

**Review by hstojic — COMMENTED** [review]
much better :)
I would only add some lay man description in the docstrings of these type vars, right now its developer oriented description

File: trieste/utils/__init__.py (hunk 4 above)
Comment (joelberkeley) [reply trieste/utils/__init__.py]: i wanted to put everything in \_\_init\_\_.py but never got round to it. Given there's nothing else in `utils`, it now may as well be in a single utils.py file (no package)

File: trieste/acquisition/function/entropy.py [hunk 5]
```diff
@@ -69,11 +67,11 @@ def __init__(
 
     @overload
     def __init__(
-        self: "MinValueEntropySearch[T]",
+        self: "MinValueEntropySearch[ProbabilisticModelType]",
```
Comment (joelberkeley) [inline trieste/acquisition/function/entropy.py:70]: curious why you've typed `self`

File: trieste/utils/__init__.py (hunk 4 above)
Comment (uri-granta) [reply trieste/utils/__init__.py:19]: No idea! I'll move it.

Comment (uri-granta) [reply trieste/utils/__init__.py]: Ah ok. I'll migrate it in a separate PR.

**Review by henrymoss — APPROVED** [review]

**Pull request merged**